tools/optimizer/simple_ast.cpp

```diff
@@ -93,7 +93,9 @@ struct StackedStack { // a stack, on the stack
         memcpy(storage, old, sizeof(T)*used);
         alloced = true;
       } else {
-        storage = (T*)realloc(storage, sizeof(T)*available);
+        T *newStorage = (T*)realloc(storage, sizeof(T)*available);
+        assert(newStorage || !storage);
```

Comment:
it's an error for `!storage` to be true. how about if we assert on `malloc` earlier up and `realloc` here returning a non-null result? can we remove the temp var here given that?

lgtm with that